src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -447,7 +447,30 @@ internal void CompleteWriter(Exception? exception)
 
         internal void AdvanceReader(in SequencePosition consumed)
         {
-            AdvanceReader(consumed, consumed);
+            // If the reader is completed
+            if (_readerCompletion.IsCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoReadingAllowed();
+            }
+
+            long examinedIndex = consumed.GetInteger();
+            BufferSegment? examinedSegment = (BufferSegment?)consumed.GetObject();
+            if (examinedSegment != null &&
+                // Avoid the lock if we're examining the entire segment, don't need to look at the last examined index in that case
+                examinedSegment.Length - examinedIndex > 0)
+            {
+                lock (SyncObj)
+                {
+                    // If the last examined index is further than the consumed pointer, let's use the last examined index
+
+                    // _lastExaminedIndex includes the RunningIndex so we remove that to calculate how many bytes we're examining
+                    examinedIndex = Math.Max(examinedIndex, _lastExaminedIndex - examinedSegment.RunningIndex);
```

Comment:
Would we consider removing the `examinedBytes < 0` check in `AdvanceReader` instead? It could lead to some weird behavior where a flush that would have completed earlier when there were more examined bytes would not complete later after the bytes become unexamined, but I'm not sure this causes any real issues.

This wouldn't be a problem if we based backpressure on consumed rather than examined bytes like we [used to](https://github.com/dotnet/runtime/issues/28845), but I definitely think we made the right call to base backpressure on examined.

Reply:
Thinking about this more, I think we may have to do this to avoid stalls if we care about solving this problem generally.

Imagine someone examined the first payload for logging or something before passing the `PipeReader` to the next `ConnectionDelegate` or what have you, and then imagine the next stage that does the final processing of the payload tries to consume it by calling `AdvanceTo(consumed)` one subsegment at a time.